pathd/path_pcep_controller.c

```diff
@@ -673,20 +752,54 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
 	ctrl_state->pcc_opts = opts;
 	return 0;
 }
-
+int pcep_thread_event_update_best_pce(struct ctrl_state *ctrl_state, int best)
+{
+	PCEP_DEBUG(" recalculating pce precedence ");
+	if (best) {
+		struct pcc_state *best_pcc_state =
+			get_pcc_state(ctrl_state, best);
+		if (best_pcc_state->previous_best != best_pcc_state->is_best) {
+			PCEP_DEBUG(" %s Resynchro best (%i) previous best (%i)",
+				   best_pcc_state->tag, best_pcc_state->id,
+				   best_pcc_state->previous_best);
+			best_pcc_state->status = PCEP_PCC_SYNCHRONIZING;
+			best_pcc_state->retry_count = 0;
+			best_pcc_state->synchronized = false;
+			pcep_thread_start_sync(ctrl_state, best_pcc_state->id);
+		} else {
+			PCEP_DEBUG(
+				" %s No Resynchro best (%i) previous best (%i)",
+				best_pcc_state->tag, best_pcc_state->id,
+				best_pcc_state->previous_best);
+		}
+	} else {
+		PCEP_DEBUG(" None best pce , all pce seem disconnected");
+	}
+	return 0;
+}
 int pcep_thread_event_update_pce_options(struct ctrl_state *ctrl_state,
 					 int pcc_id, struct pce_opts *pce_opts)
 {
+	if (!pce_opts || !ctrl_state) {
+		return 0;
+	}
 	struct pcc_state *pcc_state;
 	struct pcc_opts *pcc_opts;
 
-	if (pcc_id > ctrl_state->pcc_count) {
-		pcc_state = pcep_pcc_initialize(ctrl_state, pcc_id);
-		set_pcc_state(ctrl_state, pcc_state);
+	int current_pcc_id = get_pcc_id_by_ip_port(ctrl_state, pce_opts);
+	if (current_pcc_id) {
+		pcc_state = get_pcc_state(ctrl_state, current_pcc_id);
 	} else {
-		pcc_state = get_pcc_state(ctrl_state, pcc_id);
+		pcc_state = pcep_pcc_initialize(ctrl_state,
+						get_next_id(ctrl_state));
+		if (set_pcc_state(ctrl_state, pcc_state)) {
+			memset(pcc_state, 0, sizeof(*pcc_state));
```

Comment:
Why are you doing this ?

Reply:
For  past experiences about reusing mem without initializtion but as far as we use heare XCALLOC it's not needed.